### PR TITLE
OS Independent Theme Switcher

### PR DESCRIPTION
**Pull request by samstorment**
# Overview

Added a `<select>` to the site header that lets you choose between light, dark, and system theme. 

## Theme Select

The theme select could definitely be placed elsewhere on the page. It's always felt weird to me that so many sites prominently feature the theme toggle right at the top when it's something the user probably selects once and never touches again, but I went with it since it's pretty standard these days.

Could also opt for a toggle button with just an icon that only allows for light/dark and never displays the system option to the user, which a lot of sites seem to do.  I don't really have a preference.

I  gave the `<select>` a screen reader-only label since the label looked a little weird next to it and it felt pretty clear what the select does just by seeing the options.

## New Script

The new `theme.js` script is pretty straightforward, but it relies on being loaded early to prevent a flash of unstyled content. This is because the theme is stored in local storage and then applied to the `<html>` element as a data attribute (`data-theme='dark'`) using JavaScript. 

If we did want to move the theme `<select>` into the footer or somewhere else further down the page, the script would probably need to be split into two scripts: one to apply the correct `data-theme` attribute ASAP and another to wire up event listeners once the `<select>` is in the DOM.

## Style Changes

The media query `@media (prefers-color-scheme: dark)` can't be used because it relies on the user's OS theme and the user can change their theme dynamically via JavaScript now. The few uses of that media query were just replaced with `[data-theme='dark']` CSS selectors.

### CODE REVIEW

File: code/services-application/search-service/resources/static/search/serp.scss [hunk 1]
```diff
@@ -228,6 +231,11 @@ header {
   }
 }
 
+#theme {
```
Comment (vlofgren) [inline code/services-application/search-service/resources/static/search/serp.scss:234]: I think one thing that might be good is to have `#theme` be `display:none` unless `body[data-has-js='true']`.  This means  the experience degrades gracefully if you disable JS.

**Comment by samstorment** [comment]
Good call, I hid `#theme` when there's no JS.

In the same vein, I realized dark theme stopped working altogether when JS was disabled since it relied on `[data-theme='dark']` existing on `<html>` but that was applied via JS. I re-added `media: (prefers-color-scheme: dark)` and included the dark styles only when JS is disabled. I used a SCSS mixin so I didn't have to write the dark styles twice. 

The selector in the media query uses `:has`, which is fairly well supported at this point, but it would probably make sense to just put both `[data-theme]` and `[data-has-js]` (and any future global data attributes) on the same element whether that be `<html>` or `<body>` and just avoid `:has`. I'm thinking of moving `[data-has-js]` up to the `<html>` element, let me know if you'd prefer to keep it on the `<body>`.

**Comment by vlofgren** [comment]
One thing that's possibly a bit of a problem with the placement is how close it ends up being to the filter button on mobile.  I'm not sure what's a good way of solving that to be honest, as I do like the placement otherwise.

image

Maybe we should just remove the (ugly and misaligned) filter button, like only have the swipe action?  Or hide the manual  switcher when input is coarse?  Seems like a lesser functionality regression.

**Comment by samstorment** [comment]
I'm thinking it's minor enough to fix sometime in the future, so I decided to go for a simple quick fix for now. I just made the filter button a bit bigger so it's easier to click and aligned it. 

image

I can try out some other designs for the theme select or filter button if you think it needs to be addressed in this PR.

**Comment by vlofgren** [comment]
Making the button bigger is probably a good solution, it looks like it should be easy enough to click the right one.

**Review by vlofgren — APPROVED** [review]
Looks good.  If you feel it's ready we can merge.

**Comment by samstorment** [comment]
Yep merge away

**Pull request merged**